README
To compile:
g++ -o gah -lncurses main.cpp
OR
make


NOTE:
This readme is now very outdated, until i

New as of 1.35 Beta:
Support of terminals larger then 80x24 and to resize the terminal.


Since its beta, I suggest not testing to see if you can crash it.  Give commands exactly what they expect.
CTRL+C will execute endwin() [good].  SIGQUIT, SIGABRT, SIGKILL, SIGHUP (not an issue I guess) WON'T.
If you want to find out why you shouldnt use anything but SIGINT, try it!

gah usage:
./gah [-d]

If -d is present (actualy if argc>1), debug mode is initiated.  Every function is logged as it is entered and exited, and often data relevant to the function call is as well (such as a received or sent IM).  The log file location is gah's home dir, named "debug".   Note that this log file averages 15-20kb for short usages.  It also is of no use unless you intend to alter the code.  

After execution, gah will initilize its classes, create sockets, and resolve toc.oscar.aol.com.  If all works, it proceeds to prompt for your username and password.  Your username must be entered in a normalized (lowercase, no spaces) form.

The screen will then output one of the following:
Connecting12345678
Each number represents a step on the AIM signon process.  This is so you can identify where it went wrong.  Here is the steps:
1. Connect
2. FLAPON String
3. FLAPON Packet (recv)
4. FLAPON Packet (send)TIOCGWINSZ
5. toc_signon
6. SIGN_ON (You are now authenticated)
7. toc_add_buddy <self> (You now appear on peoples buddy lists)
8. toc_init_done

The UI will then be activated along with the main control loop.  There will immediatly be data waiting for you on the network, so a number of things will happen:
You buddy list will become populated, and toc_add_buddy sent so you recv updates of your buddies
A number of UI entries will be set


You are now in gah, the following commands will work:
NOTE: All usernames MUST be entered normalized!

/query <user>  -- start a query with the user.  The users information will be displayed above your current status, and all input not proceeded with a '/' will be sent as an IM to this person.

/msg <user> <msg> -- Send a message to a user without querying them
/auto <user> <msg> -- Same as above, except the message appears as an auto-response (dont make me regret this feature)

/away [msg] -- Set or unset away (toggles).  The away message is optional.  Note that there will be no auto-response to users who message you!

/idle HH:mm -- Set your idle time.  TOC seems to support idle times of 1 month, with the hours filed a little over 1000.  If you run it without a time, it sets your idle time as 0, and starts counting up, if your already idle, it unsets your idle time.  Being idle is cosmetic only, you can still send and recv messages while idle.

/whois <user> -- Gives you detailed information about that specific user.
/who -- Prints out everyone who is online on your buddy list and if they are available (idle/away)

/quit ...
/exit ...

Known issues:
Its easy to cause buffer overruns, if you run into a segmentation fault, please tell me where and what you were doing (debug logs are helpful).
Terminals smaller then 80x24 cause bad things.
Resizing causes weird things to happen to the query line of status window.

The UI has issues.  Such as the cursor being displayed outside of the input window.


If you need to exit immediatly:
SIGINT closes the network socket and then runs endwin.
SIGQUIT is not handled, but it will fuck up your terminal without endwin.


ChangeLog
Version 1.39

Bugs Fixed:

Known Issues:

New Features:
Primitive logging support.  It now has the ability to log all messages (in/out) and status changes (ie sign on/off, idle/away, etc) to a single log file.

Version 1.38

Bugs Fixed:
A signon issue.  Now let us never speak of this again.

Known Issues:
Same as 1.37 minus the signon issue

New Features:
If you /query someone not on your buddy list, that user will be added TEMPORARILY to it.  This means you will receive update notifications, and the status line will be filled in, but no buddy list settings are actualy altered on the server.  The user must be online to receive said information (although if you add an offline user, you will receive a message for his signon)
/quote is a new command, it supports raw commands.  If you dont know what this is or does, stay away from it.
The encrypted option in the config file has been removed.  If a password begins with "0x" it is assume to be roasted (this goes for the command line and password prompt as well).


Version 1.37

Bugs Fixed:
-

Known Issues:
Same as 1.36 Plus:
If someones online or idle time exceeds 99 hours and 59 minutes, the status window does bad things
Its currently in a non-working state for me, connection to aim is being dropped for unknown reasons.


New Features:
The following command line options have been added:
-d -- debug
-v -- version (quit after printing ver number)
-u -- specify your login name
-p -- specify your password
-P -- Since TOC works on ALL ports, specify your port to connect to here (default is 5190)
-h -- help on these commands
-c -- Specify config file

Configuration files:
A sample file is included, currently gah checks its current folder for gah.conf, then /etc.  You can also specify the location on the command line.

Version 1.36

Bugs Fixed:
Code optimisation, migrated several for loops to a function call.
holding down enter previously flooded your console with blank outgoing messages.  Since the content of these messages was null, they were never delivered, all they did was kill your message queue.
You can now query yourself.  You must be proud to have such a good friend.

Known Issues:
Lowering the vertical resolution of the terminal.  Blank lines are now inserted so everything is scrolled up, however there is an extra blank line placed in I cant seem to get rid of.
Increasing the horizontal resolution of the terminal causes the query line of the status window to only fill in blue to its previous end of line.  Seems to fix itself after a couple of minutes or so.. weird.
When increasing or decreasing the terminal verticly, there may be a blue char left at the far right side from where the status window used to be.
Do not lower the terminals width to below 60.  Do not lower the height to below 4.  These are theoretical limits, things may fuck up as you approach these.


New Features:
/wi is an alias to /whois
/whois with no arguments performs a whois on your current query
/whoami displays a whois on yourself (/whois'ing yourself works as well)
when you /query someone not on your buddy list, their nick will appear on the status window



Version 1.35

Bugs Fixed:
When your warning level decreases, you get a message saying you were just warned.  That message is now supressed.

log.h
/* log.h  Version 1.0 Beta -- Logs stuff for debug mode
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 */

class debuglog {
	private:
		bool debug;
		char *buf;
		FILE *log;
		int pid;
	public:
		debuglog(bool);
		void start(char*);
		void end();
		void write(char*);
		void fatal(char*);
		void focus(char*);
		void fwrite(char*,char*);
		void fend(char*);
		void close(char*);
};
debuglog::debuglog(bool debugtmp) {
	debug=debugtmp;
	if(debug) {
		pid=getpid();
		log=fopen("debug","a");
		fprintf(log,"%i->gah version %s\n",pid,VERSION);
		fflush(log);
	}
}
void debuglog::start(char *what) {
	buf=what;
	if(debug) {
		fprintf(log,"%i->Entering %s Function.\n",pid,what);
		fflush(log);
	}
}
void debuglog::end()  {
	if(debug) {
		fprintf(log,"%i->Exited %s Successfully.\n",pid,buf);
		fflush(log);
	}
	buf=NULL;
}
void debuglog::write(char *what) {
	if(debug) {
		fprintf(log,"%i->[%s] %s\n",pid,buf,what);
		fflush(log);
	}
}
void debuglog::fatal(char *what) {
	if(debug) {
		fprintf(log,"%i->Fatal Error Occured\n%i->[%s] %s\n",pid,pid,buf,what);
		fflush(log);
	}
}
void debuglog::focus(char *what) {
	buf=what;
}
void debuglog::fwrite(char *what,char *msg) {
	buf=what;
	write(msg);
}
void debuglog::fend(char *what) {
	buf=what;
	end();
}	
void debuglog::close(char *what) {
	if(debug) fprintf(log,"%i->[%s] %s\n%i->gah is exiting, log closed.\n",pid,buf,what,pid);
}



class logging {
	private:
		FILE *log;
		bool on;
	public:
		logging(bool);
		void im(char*,char*,bool,bool);
		void status(buddylist*);
};
logging::logging(bool ontmp) {
	on=ontmp;
	if(on) {
		log=fopen("log","at");
	}
}

void logging::im(char *who,char *msg, bool autor, bool recv) {
	if(on) {
		fprintf(log,"%s%c%s%c %s\n",(recv ? "<-":"->"),(autor ? '|':'<'),who,(autor ?'|':'>'),msg);
		fflush(log);
	}
}

void logging::status(buddylist *y) {
	if(on) {
		fprintf(log,"%s is %s, idle for %imin, %i%%warning, signed on at %i and is %s\n",y->fuser,(y->online ? "online":"offline"),y->idle/60,y->warn,y->signon,(y->avail ? "available":"away"));
		fflush(log);
	}
}


toc.h
/*
 *
 * gah -- toc.h -- v1.36 Beta
 * Performs all functions needed to connect to AIM using the TOC protocol and
 * maintain that connection.  It detects user and network input and processes
 * it on the fly.
 *
 * 
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 */

class toc {
	private:
		debuglog *log;
		logging *logger;
		struct buddylist bdlist[MAX_BUDDIES+1],*me;
		struct settings data;
		int sockfd,seq,buddies,totalbuds;
		char passwd[35],cuser[17];
		struct hostent *he;
		struct sockaddr_in dst;
		userint gui;

		/* Too Many Functions */
		void parse(char*);
		void parsenet(char*);
		void roastpass(char*);

		void upidle();
		void whois(char*);
		void whom();
		void setidle(char*);
		void setaway(char*);
		void setupbud(char*);
		int updatebud(char*);
		int findbuddy(char*);
		void normalize(char*,char*);
		void addslashes(char*,char*);

		void sndim(char*,char*,bool);
		void snddata(int,int,char*);
		void rcvdata();
	public:
		toc(debuglog*,char*);
		void setlogin(char*,char*);
		void signon();
		void loop();
};

toc::toc(debuglog *logtmp,char *configfile) {
	conf cfile(configfile,&data);
	logging loggertmp(data.log);
	logger=&loggertmp;
	cuser[0]='\0';
	me=&bdlist[MAX_BUDDIES];
	bdlist[MAX_BUDDIES].avail=true;
	strcpy(me->group,"n/a");
	gui.log=logtmp;
	log=logtmp;
	log->start("init");
	if((he=gethostbyname(data.host))==NULL) {
		cerr<<data.host<<" failed to resolve"<<endl;
		exit(1);
	}
	sockfd=socket(AF_INET,SOCK_STREAM,0);
	dst.sin_family=AF_INET;
	dst.sin_port=htons(data.port);
	dst.sin_addr=*((struct in_addr *)he->h_addr);
	memset(&dst.sin_zero,'\0',8);
	seq=0;
	log->end();
}

void toc::setlogin(char *user,char *pass) {
	log->start("setlogin");
	char *buf=pass;
	if(user==NULL) {
		if(data.username[0]=='\0') {
			cout<<"Login: ";
			cin.getline(me->nuser,16);
		}
		else {
			strcpy(me->nuser,data.username);
		}
	}
	else {
		strncpy(me->nuser,user,16);
	}
	if(pass==NULL) {
		if(data.password[0]=='\0') {
			char password[17];
			buf=password;
			struct termios n,s;
			tcgetattr(0,&s);
			n=s;
			n.c_lflag &=(~ECHO);
			cout<<"Password: ";
			tcsetattr(0,TCSANOW,&n);
			cin.getline(password,16);
			tcsetattr(0,TCSANOW,&s);
			cout.put('\n');
			roastpass(password);
		}
		else {
			buf=data.password;
		}
	}
	if(buf[0]=='0' && buf[1]=='x') {
		strcpy(passwd,buf);
	}
	else {
		roastpass(buf);
	}
	log->fend("setlogin");
}

void toc::roastpass(char *pass) {
	log->start("roastpass");
	char roast[]="Tic/Toc";
	int pos=2;
	int x;
	strcpy(passwd,"0x");
	for(x=0;(x<150) && pass[x];x++)
		pos+=sprintf(&passwd[pos],"%02x",pass[x]^roast[x%7]);
	passwd[pos]='\0';
	log->end();
}

	

void toc::loop() {
	char buf[BUFFER_SIZE];
	char *p=buf;
	int inbuf;
	fd_set a,b;
	timeval wait;
	FD_ZERO(&a);
	FD_SET(1,&a);
	FD_SET(sockfd,&a);
	wait.tv_usec=0;
	gui.start();
	while(1) {
		while(allisgood) {
			wait.tv_sec=60;
			b=a;
			select(sockfd+1,&b,NULL,NULL,&wait);
			if(FD_ISSET(STDIN,&b)) {
				inbuf=gui.getinput();
				if(isascii(inbuf)) {
					if(inbuf=='\n') {
						p[0]='\0';
						if(buf[0]=='\0') continue;//no blank lines
						gui.eol();
						parse(buf);
						p=buf;
					}
					else {
						p[0]=inbuf;
						gui.putch((char)inbuf);
						p++;
					}
				}
				else if(inbuf==KEY_BACKSPACE) {
					if(p!=buf) {
						p--;
						if(p[0]=='\t') {
							for(int i=0;i<(9-(strlen(buf)%9));i++) gui.bksp();
						}
						else gui.bksp();
						p[0]='\0';
					}
				}
				continue;
			}
			if(FD_ISSET(sockfd,&b)) {
				log->write("data on network");
				rcvdata();
				continue;
			}
			//if neither are set, then theres been no data for 60 seconds.  keepalives are sent every 50 seconds.  the connection must be dead. OR theres been an error on select run, either way were fucked.
			sprintf(buf,"\nA fatal network error has occured: %s",strerror(errno));
			log->write(buf);
			gui.print(buf);
			break;
		}
		//handle quit
		if(sig==0) break;
		if(sig==SIGWINCH) {
			sig=0;
			allisgood=true;
			gui.redraw();
			gui.setstatus(&bdlist[MAX_BUDDIES],true);
			if(cuser[0]!='\0') {
				int y=findbuddy(cuser);
				if(y==-1) gui.setnick(cuser);
				else gui.setstatus(&bdlist[y],false);
			}
			continue;
		}
		else {
			break;
		}
	}
	close(sockfd);
	gui.close();
}

void toc::upidle() {
	log->start("upidle");
	for(int i=0;i<MAX_BUDDIES+1;i++) {
		if(bdlist[i].idle>0) {
			bdlist[i].idle+=50;//increment 50 seconds.
		}
		if(!strncmp(&bdlist[i].nuser[1],&cuser[1],strlen(&cuser[1]))) {
			gui.setstatus(&bdlist[i],false);
		}
		if(i==MAX_BUDDIES) {
			gui.setstatus(&bdlist[i],true);
		}
	}
	log->end();
}

void toc::parse(char *buf) {
	log->start("parse");
	if(buf[0]=='/') {
		log->write("command parse");
		buf++;
		if(!strncmp(buf,"query",5)) {
			log->write("query");
			buf+=6;
			strcpy(cuser,buf);
			int i=0;
			i=findbuddy(cuser);
			if(i==-1) {
			       gui.setnick(cuser);
			       setupbud(cuser);
			}
			else gui.setstatus(&bdlist[i],false);
		}
		else if(!strncmp(buf,"msg",3) || !strncmp(buf,"auto",4)) {
			log->write("msg");
			bool autor;
			if(buf[0]=='a') {
				autor=true;
				buf+=5;
			}
			else {
				autor=false;
				buf+=4;
			}
			char user[17];
			int i;
			for(i=0;buf[0]!=' ' && i<16;i++,buf++) user[i]=buf[0];
			buf++;
			sndim(user,buf,autor);
		}
		else if(!strncmp(buf,"away",4)) {
			log->write("away");
			buf+=4;
			if(buf[0]==' ') buf++;
			setaway(buf);
		}
		else if(!strncmp(buf,"idle",4)) {
			log->write("idle");
			buf+=4;
			setidle(buf);
		}
		else if(!strncmp(buf,"whois",5) || !strncmp(buf,"wi",2)) {
			log->write("whois");
			if(buf[1]=='h')
				buf+=5;
			else
				buf+=2;
			if(buf[0]=='\0') {
				whois(cuser);
			}
			else {
				buf++;
				whois(buf);
			}
		}
		else if(!strncmp(buf,"whoami",6)) {
			log->write("whoami");
			whois(me->nuser);
		}
		else if(!strncmp(buf,"who",3)) {
			log->write("who");
			whom();
		}
		else if(!strncmp(buf,"quit",4) || !strncmp(buf,"exit",4)) {
			log->close("quit");
			close(sockfd);
			gui.close();
		}
		else if(!strncmp(buf,"quote",5)) {
			buf+=6;
			log->write("User is performing a raw command, pray for them:");
			log->write(buf);
			snddata(DATA,strlen(buf),buf);
		}
		else {
			log->write("errcmd");
			gui.print("Unrecognized command.");
		}
	}
	else if(cuser[0]=='\0') {
		log->write("talking to non-existant query");
		gui.print("Your not in a query with anyone!");
	}
	else {
		log->write("talking to query");
		sndim(cuser,buf,false);
	}
	log->fend("parse");
}
void toc::whois(char *who) {
	log->start("whois");
	int i=findbuddy(who);
	if(i==-1) {
		gui.print("That user is not on your buddy list!");
	}
	else {
		char buf[256];
		sprintf(buf,"%s is in group %s, is ",bdlist[i].fuser,bdlist[i].group);
		if(bdlist[i].online) {
			struct tm *ctime;
			time_t beep;
			time(&beep);
			beep-=bdlist[i].signon;
			int idle=beep/60;
			int hr=idle/60;
			idle%=60;
			ctime=localtime(&bdlist[i].signon);
			sprintf(&buf[strlen(buf)],"online, has a %i%% warning level and signed on at %02i:%02i:%02i on %02i/%02i/%02i (%i hours and %i min ago).  They are currently ",bdlist[i].warn,ctime->tm_hour,ctime->tm_min,ctime->tm_sec,ctime->tm_mon+1,ctime->tm_mday,ctime->tm_year-100,hr,idle);
			if(bdlist[i].idle>0) {
				idle=bdlist[i].idle/60;
				hr=idle/60;
				idle%=60;
				sprintf(&buf[strlen(buf)],"idled for %i hours and %i minutes",hr,idle);
			}
			else {
				strcat(buf,"not idled");
			}
			if(bdlist[i].avail) {
				strcat(buf," and are not set as away.");
			}
			else {
				strcat(buf," and are marked as away.");
			}
			log->write(buf);
			gui.print(buf);
		}
		else {
			log->write("offline");
			strcat(buf,"offline.");
			gui.print(buf);
		}
	}
	log->fend("whois");
}

void toc::whom() {//bdlist print
	log->start("whom");
	char buf[120];
	for(int x=0;x<MAX_BUDDIES;x++) {
		if(bdlist[x].online==false) continue;
		sprintf(buf,"%s from group %s is ",bdlist[x].fuser,bdlist[x].group);
		if(bdlist[x].online) {//im aware this will always be true
			strcat(buf,"online ");
			if(bdlist[x].avail) {
				strcat(buf,"and available.");
			}
			else {
				strcat(buf,"but away.");
			}
			
		}
		gui.print(buf);
	}
	log->fend("whom");
}
void toc::setidle(char *buf) {
	// HH:mm (HH can be really high)
	log->start("idle");
	int hrs,min,sec;
	if(buf[0]=='\0') {
		if(me->idle==0) sec=1;
		else sec=0;
	}
	else {
		buf++;
		hrs=atoi(buf);
		for(;buf[0]!=':' && buf[0]!='\0';buf++);
		if(buf[0]=='\0') min=0;
		else {
			buf++;
			min=atoi(buf);
		}
		min+=hrs*60;
		sec+=min*60;
	}
	char buff[80];
	sprintf(buff,"toc_set_idle %i",sec);
	snddata(DATA,strlen(buff),buff);
	log->fend("setidle");
}

void toc::setaway(char *buf) {
	log->start("setaway");
	if(me->avail) me->avail=false;
	else me->avail=true;
	gui.setstatus(&bdlist[MAX_BUDDIES],true);
	char buff[256]="toc_set_away";
	if(buf[0]!='\0') {
		sprintf(&buff[12]," \"%s\"",buf);
	}
	snddata(DATA,strlen(buff),buff);
	log->fend("setaway");
}

void toc::sndim(char *who,char *msg,bool autor) {
	log->start("sndim");
	char buf[BUFFER_SIZE];
	char whom[17];
	char mesg[BUFFER_SIZE];
	normalize(whom,who);
	addslashes(mesg,msg);
	log->write(msg);
	log->write(mesg);
	sprintf(buf,"toc_send_im %s \"%s\"",whom,mesg);
	if(autor) {
		strcat(buf," auto");
	}
	log->fwrite("sndim",buf);
	int x=findbuddy(who);
	if(x>-1) who=bdlist[x].fuser;
	gui.rcvim(who,msg,autor,true);//report the same data to be displayed as outgoing
	logger->im(who,msg,autor,false);
	snddata(DATA,strlen(buf),buf);
	log->fend("sndim");
}

void toc::signon() {//not in curses mode yet fyi
	log->start("signon");
	/* Connect */
	cout<<"Connecting";
	if(connect(sockfd,(struct sockaddr *)&dst,sizeof(struct sockaddr))==-1) {
		cerr<<"Unable to connect!"<<endl;
		exit(1);
	}
	cout.put('1');//these will be replaced with .'s for releases
	
	/* FLAPON String */
	if(send(sockfd,"FLAPON\r\n\r\n",10,0)==-1) {
		cerr<<"Failed to send FLAPON string"<<endl;
		close(sockfd);
		exit(1);
	}
	cout.put('2');

	/* FLAPON Packet-R */
	rcvdata();
	cout.put('3');

	/* FLAPON Packet-S */
	struct sodata so;
	strcpy(so.username,me->nuser);
	so.len=htons(strlen(me->nuser));
	snddata(SIGNON,8+strlen(me->nuser),(char *)&so);
	cout.put('4');
	
	/* toc_signon */
	char buf[120];
	sprintf(buf,"toc_signon login.oscar.aol.com 5190 %s %s english \"assmonkez\"",me->nuser,passwd);
	snddata(DATA,strlen(buf),buf);
	cout.put('5');

	/* SIGN_ON */
	rcvdata();
	cout.put('6');

	/* toc_add_buddy */
	sprintf(buf,"toc_add_buddy %s",me->nuser);
	snddata(DATA,strlen(buf),buf);
	cout.put('7');

	/* toc_init_done */
	strcpy(buf,"toc_init_done");
	snddata(DATA,strlen(buf),buf);
	cout.put('8');

	cout<<endl<<"Connected!"<<endl;
	log->fend("signon");

}

void toc::snddata(int type,int len,char *buf) {//len never includes null terminator here
	log->start("snddata");
	struct flaphdr sflap;
	sflap.ast='*';
	sflap.type=type;
	sflap.seq=htons(seq++);//i think this is right...
	len++;
	if(type==SIGNON && buf[0]==0) {
		len--;//this is easier then adding multiple if statements..
	}
	sflap.len=htons(len);
	char buff[len+6];
	memcpy(buff,&sflap,6);
	memcpy(&buff[6],buf,len);
	if(send(sockfd,&buff,len+6,0)==-1) {
		gui.print("A fatal network error has occured while sending data:");
		gui.print(strerror(errno));
		close(sockfd);
		gui.close();
	}
	log->fend("snddata");
}

void toc::rcvdata() {
	log->start("rcvdata");
	char buf[4097];
	struct flaphdr sflap;
	if(recv(sockfd,&sflap,6,0)==-1 || recv(sockfd,&buf,ntohs(sflap.len),0)==-1) {
		cout<<endl;
		printf("A fatal network error has occured:\n%s\n",strerror(errno));
		close(sockfd);
		if(seq>0) {
			gui.print(buf);
			gui.close();
		}
		else {
			cerr<<buf<<endl;
			exit(1);
		}
	
	}
	buf[ntohs(sflap.len)]='\0';
	if(sflap.type==DATA) {
		log->write("data msg recv");
		parsenet(buf);
	}
	else if(sflap.type==KEEPALIVE) {
		log->write("keepalive recv");
		upidle();
	}
	else if(sflap.type==SIGNON) {
		log->write("flapon recv");
		seq=ntohs(sflap.seq);
	}
	else {
		sprintf(buf,"A malformed packet has been received.",strerror(errno));
		gui.print(buf);
		close(sockfd);
		gui.close();
	}
	log->fend("rcvdata");
}

void toc::setupbud(char *user) {
	//takes formated or normalized
	char username[17];
	normalize(username,user);
	char buf[32];
	sprintf(buf,"toc_add_buddy %s",username);
	snddata(DATA,strlen(buf),buf);
}

int toc::updatebud(char *user) {
	//note that this is a TEMPORARY add, it will not be saved past this connection!
	//takes a normalized username
	log->start("updatebud");
	log->write(user);
	if(totalbuds>=MAX_BUDDIES) {
		gui.print("error: updatebud() failed: no more spaces left!");
		log->fend("updatebud");
		return -1;
	}
	else {
		strncpy(bdlist[++totalbuds].nuser,user,16);
		strncpy(bdlist[totalbuds].fuser,user,16);//until a formatted username is set, we need to do this
	}
	return totalbuds;
}


int toc::findbuddy(char *user) {// this now takes normalized usernames
	log->start("findbuddy");
	log->write(user);
	for(int i=0;i<MAX_BUDDIES+1;i++) {
		if(!strcmp(bdlist[i].nuser,user)) {
			log->end();
			return i;
		}
	}
	log->end();
	return -1;
}

void toc::normalize(char *dst,char *src) {
	log->start("normalize");
	for(;src[0]!='\0';src++,dst++) {
		if(src[0]==' ') {
			dst--;
			continue;
		}
		dst[0]=tolower(src[0]);//this should do it
	}
	dst[0]='\0';
	log->end();
}

void toc::addslashes(char *dst,char *src) {
	log->start("addslashes");
	for(;src[0]!='\0';src++,dst++) {
		if(src[0]=='$' || src[0]=='{' || src[0]=='}' || src[0]=='[' || src[0]==']' || src[0]=='(' || src[0]==')' || src[0]=='"' || src[0]=='\'' || src[0]=='\\') {
			dst[0]='\\';
			dst++;
		}
		dst[0]=src[0];
	}
	dst[0]='\0';
	log->end();
}
		

void toc::parsenet(char *buf) {//i need to break this up into functions...
	log->start("parsenet");
	if(!strncmp(buf,"IM_IN",5)) {
		//IM_IN:buddyname:F:the F is for the auto response T/F
		log->write("im recv");
		int i;
		char username[17];
		buf+=6;
		for(i=0;buf[0]!=':';i++,buf++) username[i]=buf[0];
		buf+=3;
		username[i]='\0';
		gui.rcvim(username,buf,false,false);
		logger->im(username,buf,false,true);
	}
	else if(!strncmp(buf,"UP",2)) {
		log->write("buddy update recv");
		//UPDATE_BUDDY:buddyname:T:xxx:yyy:zzz:aaa
		//xxx == evil amt (percentage)
		//yyy == signon time (epoc)
		//zzz == idle time (minutes) -- We store this in seconds!
		struct buddylist y;
		buf+=13;
		int i;
		for(i=0;buf[0]!=':';i++,buf++) y.fuser[i]=buf[0];
		buf++;
		y.fuser[i]='\0';
		normalize(y.nuser,y.fuser);
		i=findbuddy(y.nuser);
		if(i==-1) {
			i=updatebud(y.nuser);
			strcpy(bdlist[i].fuser,y.fuser);
		}
		y.online=bdlist[i].online;
		if(buf[0]=='T') {
			bdlist[i].online=true;
		}
		else {
			bdlist[i].online=false;
		}
		buf+=2;
		y.warn=bdlist[i].warn;
		bdlist[i].warn=atoi(buf);
		for(;buf[0]!=':';buf++);
		buf++;
		y.signon=bdlist[i].signon;
		bdlist[i].signon=atoi(buf);
		for(;buf[0]!=':';buf++);
		buf++;
		y.idle=bdlist[i].idle;
		bdlist[i].idle=atoi(buf)*60;
		for(;buf[0]!=':';buf++);
		buf+=2;//just changed this.. prev. val. buf+=3
		if(i==MAX_BUDDIES) buf++;
		y.avail=bdlist[i].avail;
		if(buf[0]=='U') {
			bdlist[i].avail=false;
		}
		else {
			bdlist[i].avail=true;
		}

		if(y.online!=bdlist[i].online) {
			char buffer[80];
			sprintf(buffer,"%s has just signed ",bdlist[i].fuser);
			if(bdlist[i].online) {
				strcat(buffer,"on.");
				buddies++;
			}
			else {
				strcat(buffer,"off.");
				buddies--;
			}
			gui.print(buffer);
		}
		if(!strcmp(bdlist[i].nuser,cuser)) {
			gui.setstatus(&bdlist[i],false);
		}
		if(i==MAX_BUDDIES) {
			gui.setstatus(&bdlist[i],true);
		}
		logger->status(&bdlist[i]);
	}
	else if(!strncmp(buf,"SIGN_ON",7)) {
		log->write("signon recv");
		me->online=true;
		me->avail=true;
	}
	else if(!strncmp(buf,"NICK",4)) {
		log->write("nick recv");
		buf+=5;
		strcpy(me->fuser,buf);
		me->fuser[strlen(buf)]='\0';
	}
	else if(!strncmp(buf,"CONFIG",6)) {
		log->write("config recv");
		buf+=7;
		char addbud[512]="toc_add_buddy";
		char cgroup[48]="nogroup";//cover my ass
		int y=0;
		for(totalbuds=0;buf[0]!='\0';totalbuds++,buf++) {
			if(buf[0]=='g') {
				log->write("new group");
				buf+=2;
				for(y=0;buf[0]!='\n';buf++,y++) {
					cgroup[y]=buf[0];
				}
				cgroup[y+1]='\0';
				totalbuds--;
				log->write(cgroup);
			}
			else if(buf[0]=='b') {
				log->write("new person");
				buf+=2;
				for(y=0;buf[0]!='\n';buf++,y++) {
					bdlist[totalbuds].fuser[y]=buf[0];
				}
				bdlist[totalbuds].fuser[y+1]='\0';
				strcpy(bdlist[totalbuds].group,cgroup);
				normalize(bdlist[totalbuds].nuser,bdlist[totalbuds].fuser);
				log->write(bdlist[totalbuds].nuser);
				log->focus("parsenet");
				strcat(addbud," ");
				strcat(addbud,bdlist[totalbuds].nuser);
			}
			else {
				log->write("noop");
				log->write(buf);
			}
			for(;buf[0]!='\n';buf++);
		}
		snddata(DATA,strlen(addbud),addbud);//send out the addbud so we recv updates
	}
	else if(!strncmp(buf,"ERROR",5)) {
		log->write("error");
		buf+=6;
		int error=atoi(buf);
		char buff[200];
		buf+=4;
		switch(error) {//fatal errors here are written to gui and then exited, erasing the gui... fix this
			case 901://user not avail
				sprintf(buff,"%s is not available.",buf);
				gui.print(buff);
				break;
			case 902://warning not available
				break;
			case 903:
				gui.print("A message has been dropped, you are exceeding the server speed limit!");
				break;

			case 911:
				gui.print("Error validating input");
				break;
			case 912:
				gui.print("Invalid account");
				break;
			case 913:
				gui.print("Server-side error encountered while processing request");
				break;
			case 914:
			case 981:
				gui.print("The AOL AIM service is unavailable.");
				break;

			case 950:
				break;

			case 960:
				sprintf(buff,"You are sending messages too fast to %s!",buf);
				gui.print(buff);
				break;
			case 961:
				sprintf(buff,"You missed a message from %s because it was too big.",buf);
				gui.print(buff);
				break;
			case 962:
				sprintf(buff,"You missed a message from %s because it was sent too fast.",buf);
				gui.print(buff);
				break;

			case 980:
				gui.print("Incorrect username or password.");
				break;

			case 982:
				gui.print("Your warning level is too high to sign on.");
				break;
			case 983:
				gui.print("You have been reconnecting too frequently.");
				break;
			case 989:
				sprintf(buff,"An ungodly error has occured: %s",buf);
				gui.print(buff);
				break;
		}//end switch
	}//end if error
	else if(!strncmp(buf,"EVILED",6)) {
		log->write("warned");
		buf+=7;
		int old=me->warn;
		me->warn=atoi(buf);
		gui.setstatus(&bdlist[MAX_BUDDIES],true);
		if(old<me->warn) {
			for(;buf[0]!=':';buf++);
			buf++;
			char buff[80]="You were just warned by ";
			if(buf[0]=='\0')
				strcat(buff,"anonymous");
			else
				strcat(buff,buf);
			sprintf(&buff[strlen(buff)],".  Your warning level is now %i%%.",me->warn);
			gui.print(buff);
		}
	}
	else {
		log->write("unsupported");
		gui.print("Received a TOC command not supported by gah:");
		gui.print(buf);
	}
	log->fend("parsenet");
}//end parsenet()
